### PR TITLE
[full-ci] Port Space actions from mixins to composables

### PR DESCRIPTION
**Pull request by dschmidt**
## Related Issue

- Fixes <issue_link>

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] ...

### CODE REVIEW

File: packages/web-app-admin-settings/src/components/Users/ContextActions.vue [hunk 1]
```diff
@@ -66,13 +68,24 @@ export default defineComponent({
...
+
     const menuItemsPrimaryActions = computed(() =>
       [...instance.$_edit_items, ...instance.$_delete_items].filter((item) =>
         item.isEnabled(unref(filterParams))
       )
     )
     const menuItemsSecondaryActions = computed(() =>
-      [...instance.$_editQuota_items].filter((item) => item.isEnabled(unref(filterParams)))
+      [...unref(editQuotaActions)].filter((item) =>
+        // Why does this work at all? Edit Quota is a space action and props.items is of type User[]
```
Comment (dschmidt) [inline packages/web-app-admin-settings/src/components/Users/ContextActions.vue:86]: FIXME

File: packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsDelete.spec.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        id: '1',
+        root: {
+          permissions: [{ roles: ['manager'], grantedToIdentities: [{ user: { id: 1 } }] }],
+          deleted: { state: 'trashed' }
+        }
+      }
+      const { wrapper } = getWrapper({
+        setup: ({ actions }) => {
+          // FIXME
+          // expect(unref(actions)[0].isEnabled({ resources: [buildSpace(spaceMock)] })).toBe(true)
```
Comment (dschmidt) [inline packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsDelete.spec.ts:47]: FIXME

File: packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsDisable.spec.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      const spaceMock = {
+        id: '1',
+        root: {
+          permissions: [{ roles: ['manager'], grantedToIdentities: [{ user: { id: 1 } }] }]
+        }
+      }
+      const { wrapper } = getWrapper({
+        setup: ({ actions }) => {
+          // FIXME
+          // expect(unref(actions)[0].isEnabled({ resources: [buildSpace(spaceMock)] })).toBe(true)
```
Comment (dschmidt) [inline packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsDisable.spec.ts:33]: FIXME

File: packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsEditQuota.spec.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+function getWrapper({
+  canEditSpaceQuota = false,
+  setup
+}: {
+  canEditSpaceQuota?: boolean
+  setup: (
+    instance: ReturnType<typeof useSpaceActionsEditQuota>,
+    {
+      storeOptions
+    }: // clientService
```
Comment (dschmidt) [inline packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsEditQuota.spec.ts:67]: remove comment

File: packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsEditQuota.spec.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    {
+      storeOptions: typeof defaultStoreMockOptions
+      // clientService: ReturnType<typeof defaultComponentMocks>['$clientService']
```
Comment (dschmidt) [inline packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsEditQuota.spec.ts:70]: remove comment

File: packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsEditReadmeContent.spec.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      }
+
+      const { wrapper } = getWrapper({
+        setup: ({ actions }) => {
+          // FIXME
+          // expect(
+          //   unref(actions)[0].isEnabled({
+          //     resources: [buildSpace(spaceMock)]
+          //   })
+          // ).toBe(true)
```
Comment (dschmidt) [inline packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsEditReadmeContent.spec.ts:22]: FIXME

File: packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsEditReadmeContent.spec.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+})
+
+function getWrapper({
+  setup
+}: {
+  setup: (
+    instance: ReturnType<typeof useSpaceActionsEditReadmeContent>,
+    {
+      storeOptions
+    }: // clientService
```
Comment (dschmidt) [inline packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsEditReadmeContent.spec.ts:75]: remove comment

File: packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsEditReadmeContent.spec.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    {
+      storeOptions: typeof defaultStoreMockOptions
+      // clientService: ReturnType<typeof defaultComponentMocks>['$clientService']
```
Comment (dschmidt) [inline packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsEditReadmeContent.spec.ts:78]: remove comment

File: packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsRestore.spec.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        id: '1',
+        root: {
+          permissions: [{ roles: ['manager'], grantedToIdentities: [{ user: { id: 1 } }] }],
+          deleted: { state: 'trashed' }
+        }
+      }
+      const { wrapper } = getWrapper({
+        setup: async ({ actions }, { storeOptions }) => {
+          // FIXME
+          // expect(unref(actions)[0].isEnabled({ resources: [buildSpace(spaceMock)] })).toBe(true)
```
Comment (dschmidt) [inline packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsRestore.spec.ts:47]: FIXME

**Comment by ownclouders** [comment]
Results for <strong>e2e-tests oC10</strong> https://drone.owncloud.com/owncloud/web/33922/11/1
<details><summary>:boom: To see the trace, please open the link in the console ...</summary>

<p>


#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/33922/tracing/alice-can-share-this-weeks-meal-plan-with-all-parents-alice-2023-3-21-11-51-30.zip 

#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/33922/tracing/alice-can-share-this-weeks-meal-plan-with-all-parents-brian-2023-3-21-11-51-59.zip 

#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/33922/tracing/alice-can-share-this-weeks-meal-plan-with-all-parents-carol-2023-3-21-11-52-06.zip 


</p></details>
:boom: The e2e-oc10 tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>e2e-tests oCIS</strong> https://drone.owncloud.com/owncloud/web/33925/12/1
<details><summary>:boom: To see the trace, please open the link in the console ...</summary>

<p>


#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/33925/tracing/admin-user-can-change-personal-quotas-for-users-admin-2023-3-21-01-03-32.zip 

#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/33925/tracing/admin-user-can-change-personal-quotas-for-users-alice-2023-3-21-01-03-27.zip 

#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/33925/tracing/admin-user-can-change-personal-quotas-for-users-brian-2023-3-21-01-03-30.zip 

#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/33925/tracing/multiple-spaces-can-be-managed-at-once-in-the-admin-settings-via-the-batch-actions-alice-2023-3-21-01-02-29.zip 


</p></details>

File: packages/web-app-admin-settings/src/components/Users/ContextActions.vue [hunk 10]
```diff
@@ -13,7 +13,7 @@
 </template>
 
 <script lang="ts">
-import ShowDetails from '../../mixins/showDetails'
+import { useActionsShowDetails } from '../../../../web-pkg/src/composables/actions'
```
Comment (JammingBen) [inline packages/web-app-admin-settings/src/components/Users/ContextActions.vue:16]: Relative import from `web-pkg` 🙈 In general we can combine the imports from `web-pkg/src/composables/actions` into one here.

File: packages/web-pkg/src/composables/actions/spaces/useSpaceActionsEditQuota.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+import { User } from 'web-client/src/generated'
+import { useAbility } from '../../ability'
+import { useClientService, useRoute } from 'web-pkg/src/composables'
+
+export const useSpaceActionsEditQuota = ({ store }: { store?: Store<any> } = {}) => {
+  store = store || useStore()
+  const { $gettext } = useGettext()
+  const ability = useAbility()
+  const clientService = useClientService()
+  const route = useRoute()
```
Comment (JammingBen) [inline packages/web-pkg/src/composables/actions/spaces/useSpaceActionsEditQuota.ts:16]: Some unused variables here: `store`, `clientService`, `route`

File: packages/web-test-helpers/src/defaultPlugins.ts [hunk 12]
```diff
@@ -2,12 +2,13 @@ import DesignSystem from '@ownclouders/design-system'
...
 import { abilitiesPlugin } from '@casl/vue'
-import { createMongoAbility } from '@casl/ability'
+import { SubjectRawRule, createMongoAbility } from '@casl/ability'
+import { AbilityActions, AbilitySubjects } from 'web-pkg/src'
 
 export interface DefaultPluginsOptions {
   designSystem?: boolean
   gettext?: boolean
-  abilities?: any
+  abilities?: SubjectRawRule<AbilityActions, AbilitySubjects, any>[]
```
Comment (JammingBen) [inline packages/web-test-helpers/src/defaultPlugins.ts:11]:
```suggestion
  abilities?: AbilityRule[]
```

File: packages/web-test-helpers/src/helpers.ts [hunk 13]
```diff
@@ -52,3 +62,11 @@ export const createRouter = (options?: Partial<RouterOptions>) =>
...
     ...options
   })
+
+export const createMockActionComposables = (module) => {
+  const mockModule: Record<string, any> = {}
+  for (const m of Object.keys(module)) {
+    mockModule[m] = jest.fn(() => ({ actions: ref([]) }))
+  }
+  return mockModule
+}
```
Comment (JammingBen) [inline packages/web-test-helpers/src/helpers.ts:72]: We're not using that one right now, should we remove it then?

**Review by JammingBen — COMMENTED** [review]
Nice, only some nitpicks from my side. Some potential follow-ups:

* Use `LoadingService` for space actions: `rename`, `disable`, `enable`, `delete`
* Refactor the class names of the actions. Those are quite random sometimes (e.g. `oc-files-...` for space actions)

File: packages/web-pkg/src/composables/actions/useActionsShowDetails.ts [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    {
+      name: 'show-details',
+      icon: 'information',
+      label: () => $gettext('Details'),
+      handler: () => eventBus.publish(SideBarEventTopics.open),
+      isEnabled: ({ resources }) => {
+        return (resources as unknown[]).length > 0
+      },
+      componentType: 'button',
+      class: 'oc-admin-settings-show-details-trigger'
```
Comment (kulmann) [inline packages/web-pkg/src/composables/actions/useActionsShowDetails.ts:20]:
```suggestion
      class: 'oc-show-details-trigger'
```

or 

```suggestion
      class: 'oc-generic-show-details-trigger'
```

or something similar? composable is (gladly) generic and not part of the admin settings app anymore.

Same for other action composables in this PR. Can be done in a followup. Also needs adjustment of element selectors in tests.

File: packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsDelete.spec.ts [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+  defaultStoreMockOptions,
+  RouteLocation,
+  getComposableWrapper
+} from 'web-test-helpers'
+import { mock } from 'jest-mock-extended'
+import { unref } from 'vue'
+
+describe('delete', () => {
+  describe('isEnabled property', () => {
+    it('should be false when not resource given', () => {
```
Comment (kulmann) [inline packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsDelete.spec.ts:16]:
```suggestion
    it('should be false when no resource given', () => {
```

File: packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsDelete.spec.ts [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+            resources: [mock<SpaceResource>({ id: 1, canBeDeleted: () => false })]
+          })
+
+          expect(storeOptions.actions.createModal).toHaveBeenCalledTimes(0)
+        }
+      })
+    })
+  })
+
+  describe('method "$_delete_deleteSpace"', () => {
```
Comment (kulmann) [inline packages/web-pkg/tests/unit/composables/actions/spaces/useSpaceActionsDelete.spec.ts:91]: `$_delete_deleteSpace` <-- huh? 🙈 

Please also check in other tests.

**Review by kulmann — APPROVED** [review]
nice one! 💪 

some followup-nitpick-stuff. but one e2e test is failing so you can't merge anyway. Maybe you squeeze some more of the nitpick-fixes in. 🙈

File: packages/web-pkg/src/composables/actions/useActionsShowDetails.ts (hunk 14 above)
Comment (dschmidt) [reply packages/web-pkg/src/composables/actions/useActionsShowDetails.ts:20]: uh, yeah - let's do that in a followup

**Pull request merged**